src/hotspot/cpu/x86/x86.ad

```diff
@@ -8428,6 +8429,59 @@ instruct vmaskcast(vec dst) %{
   ins_pipe(empty);
 %}
 
+instruct vmaskcast_avx(vec dst, vec src) %{
+  predicate(Matcher::vector_length(n) == Matcher::vector_length(n->in(1)) &&
+            Matcher::vector_length_in_bytes(n) != Matcher::vector_length_in_bytes(n->in(1)) &&
+            (UseAVX >= 2 ||
+             (Matcher::vector_length_in_bytes(n) != 32 &&
+              Matcher::vector_length_in_bytes(n->in(1)) != 32)));
+  match(Set dst (VectorMaskCast src));
+  format %{ "vector_mask_cast $dst, $src" %}
+  ins_encode %{
+    int vlen = Matcher::vector_length(this);
+    BasicType src_bt = Matcher::vector_element_basic_type(this, $src);
+    BasicType dst_bt = Matcher::vector_element_basic_type(this);
+    __ vector_mask_cast($dst$$XMMRegister, $src$$XMMRegister, dst_bt, src_bt, vlen);
+  %}
+  ins_pipe(pipe_slow);
+%}
+
+instruct vmaskcast_avx1_32B_expand(vec dst, vec src, vec xtmp1, vec xtmp2) %{
+  predicate(Matcher::vector_length(n) == Matcher::vector_length(n->in(1)) &&
+            Matcher::vector_length_in_bytes(n) > Matcher::vector_length_in_bytes(n->in(1)) &&
+            UseAVX == 1 &&
```

Comment:
Since most x86 would be avx > 1, I would suggest testing `UseAVX == 1` first and then others.

Reply:
Will fix this in a separate patch, thanks!

Reply:
I think the general cases should come before the particular ones. These 2 rules are for a corner case where float or double masks with 256-bit width are involved. Thanks.

Reply:
> I think the general cases should come before the particular ones. These 2 rules are for a corner case where float or double masks with 256-bit width are involved. Thanks.

For modern x86 platforms, UseAVX would be >= 2, these 2 rules can be filtered out as soon as possible if we test  `UseAVX == 1` at the beginning.